openstack_controller/tests/test_openstack_sdk_api.py

```diff
@@ -0,0 +1,491 @@
+        return EXAMPLE_AGGREGATES_VALUE
+
+    def list_networks(self):
+        return EXAMPLE_NETWORKS_VALUE
+
+    def search_flavors(self, filters):
+        return EXAMPLE_FLAVORS_VALUE
+
+
+def test_get_endpoint():
+    api = OpenstackSdkApi(None)
+
+    with pytest.raises(AuthenticationNeeded):
+        api._check_authentication()
+
+    api.connection = MockOpenstackConnection()
+
+    assert api.get_keystone_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/a536052eba574bd4baf89ff83e3a23db'
+    assert api.get_nova_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/0adb9d108440437fa6841d31a989ed89'
+    assert api.get_neutron_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/408fbfd00abf4bd1a71044f4849abf66'
+
+    # Test cache
+    assert api.get_keystone_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/a536052eba574bd4baf89ff83e3a23db'
+    assert api.get_nova_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/0adb9d108440437fa6841d31a989ed89'
+    assert api.get_neutron_endpoint() == u'http://10.0.3.44:5000/v3/endpoints/408fbfd00abf4bd1a71044f4849abf66'
+
+    with mock.patch('datadog_checks.openstack_controller.api.OpenstackSdkApi._get_service',
+                    return_value={u'id': 'invalid_id'}):
+        api.endpoints = {}
+        with pytest.raises(KeystoneUnreachable):
+            api.get_keystone_endpoint()
+
+        with pytest.raises(MissingNovaEndpoint):
+            api.get_nova_endpoint()
+
+        with pytest.raises(MissingNeutronEndpoint):
+            api.get_neutron_endpoint()
+
+
+def test_get_project():
+    api = OpenstackSdkApi(None)
+    api.connection = MockOpenstackConnection()
+
+    assert api.get_projects() == EXAMPLE_PROJECTS_VALUE
+
+
+def test_get_project_limit():
+    api = OpenstackSdkApi(None)
+    api.connection = MockOpenstackConnection()
+
+    assert api.get_project_limits(u'680031a39ce040e1b81289ea8c73fb11') == {
+                "maxImageMeta": 128,
+                "maxPersonality": 5,
+                "maxPersonalitySize": 10240,
+                "maxSecurityGroupRules": 20,
+                "maxSecurityGroups": 10,
+                "maxServerMeta": 128,
+                "maxTotalCores": 20,
+                "maxTotalFloatingIps": 10,
+                "maxTotalInstances": 10,
+                "maxTotalKeypairs": 100,
+                "maxTotalRAMSize": 51200,
+                "maxServerGroups": 10,
+                "maxServerGroupMembers": 10,
+                "totalCoresUsed": 0,
+                "totalInstancesUsed": 0,
+                "totalRAMUsed": 0,
+                "totalSecurityGroupsUsed": 0,
+                "totalFloatingIpsUsed": 0,
+                "totalServerGroupsUsed": 0
+            }
+    with pytest.raises(SDKException):
+        api.get_project_limits('invalid_id')
+
+
+def test_get_os_hypervisors_detail():
+    api = OpenstackSdkApi(None)
+    api.connection = MockOpenstackConnection()
+
+    assert api.get_os_hypervisors_detail() == EXAMPLE_HYPERVISORS_VALUE
```

Comment:
For all these methods, is the output the same as what the `simpleApi` Would return.
It would be great to use the same assertion values in both cases just to make sure things match.

Reply:
Some of these methods return more information than `SimpleApi`, and some less information (not used by the check). There are some commentary for the methods that have less information.

Reply:
I guess the tests then should only assert for values actually used by the check, then we can keep the same assertions for both APIs